=htc #17403 fix connection pool reusing existing connections too early

sirthias opened the pull request:
Fixes #17403

/cc @jrudolph

sirthias committed:
=htc #17403 fix connection pool reusing existing connections too early

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -67,10 +72,15 @@ private object PoolConductor {
       val doubler = Flow[SwitchCommand].mapConcat(x ⇒ x :: x :: Nil) // work-around for https://github.com/akka/akka/issues/17004
       val route = b.add(new Route(slotCount))
       val retrySplit = b.add(new RetrySplit())
+      val flatten = Flow[RawSlotEvent].mapAsyncUnordered(slotCount) {
```
Is the use of unordered correct here?

drewhk replied:
or there can be at most one Event in circulation per slot?

sirthias replied:
Unordered is fine. We don't want a slow response on one connection block us from seeing an update from another connection.

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -22,15 +24,16 @@ private object PoolSlot {
-    final case class RequestCompleted(slotIx: Int) extends SimpleSlotEvent
-    final case class Disconnected(slotIx: Int, failedRequests: Int) extends SimpleSlotEvent
-    final case class RetryRequest(rc: RequestContext) extends SlotEvent
+    final case class WhenRequestCompleted(future: Future[RequestCompleted]) extends RawSlotEvent
```
the name is a bit misleading. You can just call it RequestCompletedFuture

sirthias replied:
Ok, will do.

drewhk commented:
LGTM, if mapAsyncUnordered is correct here.

drewhk commented:
@sirthias can I merge this? Is mapAsyncUnordered OK?

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -149,8 +152,21 @@ private object PoolSlot {
+            x.copy(data = newData) -> whenCompleted
+          case x: HttpEntity.Chunked ⇒
+            val (newChunks, whenCompleted) = StreamUtils.captureTermination(x.chunks)
+            x.copy(chunks = newChunks) -> whenCompleted
```
This does not handle all entity types: what about IndefiniteLength?

jrudolph replied:
`IndefiniteLength` is not a `ResponseEntity`.

sirthias replied:
`HttpEntity.IndefiniteLength` is not a `ResponseEntity`. Therefore we won't see it here.

rkuhn replied:
I am massively confused by the type hierarchy here, it seems, but I trust you.

sirthias replied:
It's all really simple... :)
4 super types (`RequestEntity`, `ResponseEntity`, `BodyPartEntity` and `UniversalEntity`) "select" sub sets of the 4 actual implementations: `Strict`, `Default`, `CloseDelimited` and `IndefiniteLength`.
Maybe we should provide a handy matrix in the docs that shows with super types comprises which implementation.

sirthias replied:
Actually its five implementations, `Chunked` was missing from my list above.
We definitely need this matrix.... :)

rkuhn replied:
My problem when reading the code on github was that it is not at all obvious how Chunked extends any of RequestEntity or ResponseEntity, but I might just have missed something.

rkuhn commented:
LGTM (modulo the point Endre raised about the mapAsyncUnordered—I have a hunch that it is fine but there may be dirty details there)

sirthias commented:
Ok, I renamed `WhenRequestCompleted` as suggested by @drewhk and rebased onto the current release-2.3-dev.
AFAICS there really is no issue with the `mapAsyncUnordered`. The responses from a single connection are naturally ordered and there is no connection between connections (as funny as that may sound). And even if a future for one connection should somehow overtake a previous future for the same connection then this wouldn't be a problem.

So, relax and smell the roses. :)

drewhk commented:
Wonderful :) Will merge after validation.

rkuhn commented:
Good point: the roses do indeed smell wonderful out in the garden :-)
